02_bubble/jacklee/main.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprint(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(origin []int) []int {
+	result := make([]int, len(origin))
+	copy(result, origin)
+	arrLen := len(origin)
+	for i := 0; i < arrLen-1; i++ {
+		for j := 0; j < arrLen-1-i; j++ {
+			if result[j] > result[j+1] {
+				result[j+1], result[j] = result[j], result[j+1]
+			}
+		}
+	}
+	return result
+}
+
+func heapSort(origin []int) []int {
+	heap := make([]int, len(origin))
+	copy(heap, origin)
+	makeHeap(heap)
+	result := make([]int, len(origin))
+	heapLen := len(origin)
+	for i := 0; i < heapLen; i++ {
+		swap(0, len(heap)-1, heap)
+		down(0, len(heap)-1, heap)
+		n := len(heap)
+		x := heap[n-1]
+		heap = heap[0 : n-1]
+		result[i] = x
+	}
+	return result
+}
+
+func makeHeap(origin []int) {
+	n := len(origin)
+	for i := n/2 - 1; i >= 0; i-- {
```

Comment:
It feels like this should be `(n - 1)/2`. I could be wrong, though. I haven't drilled down to check.